pulsar-client/src/main/java/org/apache/pulsar/client/impl/ZeroQueueConsumerImpl.java

```diff
@@ -84,11 +84,20 @@ protected Message<T> internalReceive() throws PulsarClientException {
 
     @Override
     protected CompletableFuture<Message<T>> internalReceiveAsync() {
+        // Just being cautious
+        if (incomingMessages.size() > 0) {
+            log.error("The incoming message queue should never be greater than 0 when Queue size is 0");
+            incomingMessages.forEach(Message::release);
+            incomingMessages.clear();
+        }
         CompletableFuture<Message<T>> future = super.internalReceiveAsync();
+        waitingOnReceiveForZeroQueueSize = true;
```

Comment:
It's confusing that we don't use `internalReceiveAsync()` to implement `internalReceive()`